usb/esp_tinyusb/tinyusb_net.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * SPDX-FileCopyrightText: 2023 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "freertos/FreeRTOS.h"
+#include "freertos/event_groups.h"
+#include "tinyusb_net.h"
+#include "descriptors_control.h"
+#include "usb_descriptors.h"
+#include "device/usbd_pvt.h"
+#include "esp_check.h"
+
+#define MAC_ADDR_LEN 6
+
+typedef struct packet {
+    void *buffer;
+    void *buff_free_arg;
+    uint16_t len;
+    esp_err_t result;
+} packet_t;
+
+struct tinyusb_net_handle {
+    bool initialized;
+    SemaphoreHandle_t buffer_sema;
+    EventGroupHandle_t  tx_flags;
+    tusb_net_rx_cb_t    rx_cb;
+    tusb_net_free_tx_cb_t tx_buff_free_cb;
+    tusb_net_init_cb_t init_cb;
+    char mac_str[2 * MAC_ADDR_LEN + 1];
+    void *ctx;
+    packet_t *packet_to_send;
+};
+
+const static int TX_FINISHED_BIT = BIT0;
+static struct tinyusb_net_handle s_net_obj = { };
+static const char *TAG = "tusb_net";
+
+static void do_send_sync(void *ctx)
+{
+    (void) ctx;
+    if (xSemaphoreTake(s_net_obj.buffer_sema, 0) != pdTRUE || s_net_obj.packet_to_send == NULL) {
+        return;
+    }
+
+    packet_t *packet = s_net_obj.packet_to_send;
+    if (tud_network_can_xmit(packet->len)) {
```

Comment:
Question: Did the function ever returned 'false' during your tests?

If yes, could we make the error handling a little more thorough?

Reply:
Yes, this happens if the USB cannot accept more packets (i.e. the available space in the outgoing buffer is less that size of this packet).

> could we make the error handling a little more thorough?

The error should be reported to the user API, as this is a synchronous API which waits for completion.